### PR TITLE
Refactor log harvester to simplify it.

### PR DESCRIPTION
**Pull request by ruflin**
Fix issue that for some encoding bytes count instead of rune count was used for offest which lead
to restart the file reading from the beginning.

### CODE REVIEW

File: harvester/log.go [hunk 1]
```diff
@@ -337,14 +338,17 @@ func readLine(reader *bufio.Reader, buffer *bytes.Buffer, partialLineWaiting tim
...
 		// If we got a full line, return the whole line without the EOL chars (LF or CRLF)
 		if !isPartialLine {
-			// Get the str length with the EOL chars (LF or CRLF)
-			bufferSize := buffer.Len()
-			str := new(string)
-			// Remove line endings
-			*str = buffer.String()[:bufferSize-lineEndingChars(segment)]
+
+			str := buffer.String()
+			// Rune count is relevant as Size() returns also rune count
```
Comment (urso) [inline harvester/log.go]: which Size() returns rune count?

**Comment by urso** [comment]
I'm feeling kinda uneasy to have offsets based on rune-count. Can we split up the harvester in some kind of input processor taking care of counting by bytes (offsets) read throught encoder + harvester using this input/file processor?

Does go have ftell like support?

File: harvester/log.go (hunk 1 above)
Comment (ruflin) [inline harvester/log.go]: I was talking about FileInfo.Size(). Will clarify it in the docs.

**Comment by ruflin** [comment]
@urso Also not too happy about this implementation because of the encoding. One option would be to do the encoding conversion after the reading so we would always read single bytes.

It seems like the Seek() could be used similar to ftell: http://stackoverflow.com/questions/10901351/fgetpos-available-in-go-want-to-find-file-position

Didn't test it yet.

**Comment by ruflin** [comment]
@urso I suggest to move forward here as it seems to solve the current issue but keep iterating on top of it.

**Comment by urso** [comment]
tests still failing?

**Comment by tsg** [comment]
@urso we're working on it, see the discussion on chat :-)

**Comment by tsg** [comment]
@ruflin, to proceed here, should we take the harvester refactoring and leave the offset fix for a different PR?

**Comment by ruflin** [comment]
Sounds good to me. The question is which one of the broken offset implementation should we take? The main problem with the initial on is that it starts to read the file from the beginning again quite often.

**Comment by ruflin** [comment]
This change will heavily conflict with https://github.com/elastic/filebeat/pull/224. Some changes will not be even required anymore. I suggest to first get https://github.com/elastic/filebeat/pull/224 in and then reapply the changes that are needed. This probably then can even be combined with https://github.com/elastic/filebeat/pull/222 

@urso FYI

**Comment by ruflin** [comment]
Closing in favor of https://github.com/elastic/filebeat/pull/232

**Pull request closed without merging**